File: two_pointers/11_container_with_most_water.cc
#include <algorithm>
#include <vector>

// 2024/01/03: 自己只写出暴力解法, 会超时.

// 思路不容易想到, 实现简单.
int MaxArea(std::vector<int>& height) {
  int left = 0, right = height.size() - 1;
  int res = 0;
  while (left < right) {
    res = std::max(res, (right - left) * std::min(height[left], height[right]));
    if (height[left] < height[right]) {
      ++left;
    } else {
      --right;
    }
  }
  return res;
}

File: two_pointers/11_container_with_most_water.py
from typing import List

# 2026/01/11: 自己写出来.
def max_area(height: List[int]) -> int:
    left = 0
    right = len(height) - 1
    result = 0

    while left < right:
        area = (right - left) * min(height[left], height[right])
        result = max(result, area)
        if (height[left] <= height[right]):
            left += 1
        else:
            right -= 1

    return result

height = [1,8,6,2,5,4,8,3,7]
print(max_area(height))
